Improve documentation after defining API (#12)
kyma-incubator/octopus: docs/06-00-CRD-TestDefinition.md

> ---

Reviewer: Metadata section is only needed for the docs in the `kyma/docs` repository as it helps to render them properly on `kyma-project.io`. There is no need for metadata section in places other than `kyma/docs`.

This applies also to file naming convention. You can go only with `tutorial`, `cr-testdefinition` and `cr-clustertestsuite` without numerals.

Aside from that, I am more than happy to see that you've read through the guidelines and templates ☺️

title: TestDefinition
type: Custom Resource
---

The `TestDefinition` CustomResourceDefinition (CRD) is a detailed description of the kind of data and the format used to define acceptance tests executed as a Pod 
on a Kubernes cluster. 
To get the up-to-date CRD and show the output in the `yaml` format, run this command:

```
kubectl get crd TestDefinition -o yaml
```

## Sample custom resource

> This is a sample resource that defines the simplest test definition. 

Reviewer: ```suggestion
This is a sample resource that provides a simple test definition. 
```

```
apiVersion: testing.kyma-project.io/v1alpha1
kind: TestDefinition
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: test-dummy
spec:
  template:
    spec:
      containers:
        - name: test
          image: alpine:latest
          command:
            - "pwd"
```

## Custom resource parameters

This table lists all the possible parameters of a given resource together with their descriptions:

| Parameter   |      Mandatory      |  Description |
|:-----------:|:-------------------:|:-------------|
| **metadata.name** |    **YES**   | Specifies the name of the CR. |

> | **spec.template** |    **YES**   | An object that describes the pod that will be created. This field is of PodTemplateSpec type from Kubernetes API and its detailed description can be found [here](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#podtemplatespec-v1-core)  |

Reviewer: ```suggestion
| **spec.template** |    **YES**   | Describes the Pod that will be created. This field is of `PodTemplateSpec` type from the Kubernetes API. Find its detailed description [here](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#podtemplatespec-v1-core)  |
```

> | **spec.skip**     |    **NO**    | Mark that test should not be executed. The default vaule is `false`. This feature is not yet implemented. |

Reviewer: ```suggestion
| **spec.skip**     |    **NO**    | Indicates that a test should not be executed. The default value is `false`. This feature is not yet implemented. |
```

Reviewer: Should we document features that are not yet implemented?

> | **spec.disableConcurrency** | **NO** | Disallow running the test concurrently with other tests. The default vaule is `false`. This feature is not yet implemented. 

Reviewer: ```suggestion
| **spec.disableConcurrency** | **NO** | Disallows running tests concurrently. The default value is `false`. This feature is not yet implemented. 
```

Author: but in this case we do not allow to run that specific test, so IMO it cannot be `tests`. When we talking about current test, should we use `the`?

> | **spec.timeout** | **NO** | Define maximal duration of the test, after which it should be terminated and it's execution marked as a **Failed**. No The default vaule. This feature is not yet implemented. 

Reviewer: ```suggestion
| **spec.timeout** | **NO** | Defines the maximal duration of a test, after which it is terminated and marked as **Failed**. This feature is not yet implemented. 
```

## Related resources and components

These are the resources related to this CR:

| Custom resource |   Description |
|:----------:|:------|

> | ClusterTestSuite |  ClusterTestSuite defines which `TestDefinitions` to execute.  |

Reviewer: ```suggestion
| ClusterTestSuite |  Defines TestDefinitions to execute.  |
```

Reviewer: Please provide a link to this CR.

Author: It should be in the documentation that I should add link!!! :)

These components use this CR:

| Component   |   Description |
|:----------:|:------|

> | Octopus |  When a `ClusterTestSuite` is created, `Octopus` fetches `TestDefinitions` that matches the  `ClusterTestSuite` selectors.  |

Reviewer: ```suggestion
| Octopus |  When a ClusterTestSuite is created, Octopus fetches TestDefinitions that match the ClusterTestSuite selectors.  |
```
